Debug: major theming improvements 🎨 🎉

robertrossmann opened the pull request:
> Let's be honest - the debug panel is past due some major repaint. Many parts of the UI have hard-coded colours, which makes the debug UI stand out of the other parts of VS Code, in a not very pleasant way. It's time to change that!

This PR introduces several new colour contribution points for many parts of the Debug UI. The vast majority revolves around the stack trace viewlet and the various colours used for showing variables across the debug panel, and the debug REPL.

Here is the list of new colour contributions. Please be aware that the screenshots below contain colours I overrode while working on this, they dot not represent the default colours in this PR. ⚠️ I kept the original color values as defaults for all the new colour contributions in vast majority of situations. There might be a slight deviation here and there to fix contrast but the overal look and feel should be relatively the same.

Looking forward to hear your feedback! 👀

### Stack trace viewlet

`debugView.exceptionLabelForeground`
`debugView.exceptionLabelBackground`

![exception label](https://user-images.githubusercontent.com/3058150/78947033-9af2bf00-7ac4-11ea-9447-9cef49b59d8c.png)

`debugView.stateLabelForeground`
`debugView.stateLabelBackground`

![state label](https://user-images.githubusercontent.com/3058150/78947104-c8d80380-7ac4-11ea-84af-e4698c76b994.png)

`debugView.valueChangedHighlight`

> This actually "blinks" when a value changes. This is merely the final colour.
![value changed background](https://user-images.githubusercontent.com/3058150/78947145-e60cd200-7ac4-11ea-9324-84483d6238a4.png)

### Debug REPL

~`debugConsole.warningForeground`~ <- **Removed as per review**

![repl warning](https://user-images.githubusercontent.com/3058150/78947456-ac889680-7ac5-11ea-83cd-d797262a8eda.png)

~`debugConsole.annotationForeground`~ <- **Removed as per review**

> No screenshot here, not found in the UI. 🤷‍♂️ `<imagine popular Travolta gif here>`

### Variables / tokens & values

> These are the colours used in the _Variables_ viewlet and in the hovers during a debugging session.

`debugTokenExpression.name`
`debugTokenExpression.value`
`debugTokenExpression.string`
`debugTokenExpression.boolean`
`debugTokenExpression.number`

![variables](https://user-images.githubusercontent.com/3058150/78947701-63851200-7ac6-11ea-9b47-96052cde30f4.png)
![hover](https://user-images.githubusercontent.com/3058150/78947702-64b63f00-7ac6-11ea-8b9b-b4d44ce26542.png)

`debugTokenExpression.error`

![repl error](https://user-images.githubusercontent.com/3058150/78947597-17d26880-7ac6-11ea-9669-b6fd937adeaf.png)

### Other changes

- The line/column position info in the stack trace viewlet now uses the same colours as a regular "badge" (`badge.foreground` & `badge.background`)

- The filename in the stack trace viewlet now uses the same colour as for list's deemphasized content (`list.deemphasizedForeground`)
- The debug REPL's input box border now uses the input border colour (`input.border`) if defined, otherwise it defaults back to the original colour used before this PR.

![call stack](https://user-images.githubusercontent.com/3058150/78947792-ba8ae700-7ac6-11ea-8872-295b024b3ffd.png)

### Overall look and feel

| Before | After |
|:------:|:------:|
|![[before][before]][before]|![[after][after]][after]|

[before]: https://user-images.githubusercontent.com/3058150/78948464-b9f35000-7ac8-11ea-9b1e-08c9ae90ee57.png
[after]: https://user-images.githubusercontent.com/3058150/78948470-be1f6d80-7ac8-11ea-9a34-c16c1bf869dc.png

isidorn commented:
@robertrossmann thanks a lot for this awesome PR, overall great work, I really like it!

Feedback:
* debug toolbar is the debug floating widget with the pause / step action, so you are using a wrong name.  You should use probably a `debugView` prefix, or just a `debug` prefix, not the `debugToolbar`
* Can you please list all the colors you have added since we will have to name them correctly and @aeschli can help here. Also list the comments you did which each color please.
* Did you verify that your changes do not break debug coloring across: debug views, debug console, debug hover?
* Did you verify that the default coloring is left as it was for both light, dark and high contrast theme?

Also assigning to @misolori for designer review and @aeschli for review of color names.
I will review the code changes which look good from a bird's eye persepctive.

fyi @connor4312 @weinand @roblourens

robertrossmann commented:
@isidorn Thank you for your feedback! I will address it now. 💪 

- I went ahead and renamed the `debugToolBar.*` contributions to `debugView.*`. They are also updated in the original PR description. ✅ 
- I have verified that the colours work consistently in all places I regularly use in my standard debugging workflow. I checked the following:
  - all the "sub-views" in the debug panel
  - hovers which show token contents during debugging
  - the REPL console
- I have copied the original, hard-coded colours from all three theme variants - dark, light and HC. However, I did make some changes to the originals in a few places, to make them look better. Below I have prepared screenshots showing the differences in colouring I have made. The screenshots have the exact same contents so it should be easy to compare them side-by-side.

> Can you please list all the colors you have added since we will have to name them correctly...

I am not sure if I understand this correctly, but I did not really add any _new_ colours, I merely took the existing hard-coded colours and made them themeable. The colour contributions are listed in the original description, including a screenshot of the thing they apply to. For completness' sake, here they are listed again in full.

`debugView.exceptionLabelForeground`
`debugView.exceptionLabelBackground`
`debugView.stateLabelForeground`
`debugView.stateLabelBackground`
`debugView.valueChangedHighlight`
`debugConsole.warningForeground`
`debugConsole.annotationForeground`
`debugTokenExpression.name`
`debugTokenExpression.value`
`debugTokenExpression.string`
`debugTokenExpression.boolean`
`debugTokenExpression.number`
`debugTokenExpression.error`

| Theme | Original | Updated |
|:-----:|:--------:|:-------:|
|Remedy|![[remedy old]][remedy old]|![[remedy new]][remedy new]|
|Dark+|![[dark+ old]][dark+ old]|![[dark+ new]][dark+ new]|
|Light+|![[light+ old]][light+ old]|![[light+ new]][light+ new]|
|HC|![[hc old]][hc old]|![[hc new]][hc new]|

To summarise which colours are now different in writing:

- In call stack view, the filename and line/col position info
- In variables view, the variables' contents

[remedy old]: https://user-images.githubusercontent.com/3058150/78995556-eba8fd00-7b42-11ea-813f-d3769825251a.png
[dark+ old]: https://user-images.githubusercontent.com/3058150/78995562-efd51a80-7b42-11ea-9fe5-2796063084c4.png
[light+ old]: https://user-images.githubusercontent.com/3058150/78995563-f06db100-7b42-11ea-898a-5b33bd9e7a19.png
[hc old]: https://user-images.githubusercontent.com/3058150/78995566-f1064780-7b42-11ea-8344-c06a7221ead2.png
[dark+ new]: https://user-images.githubusercontent.com/3058150/78995568-f19ede00-7b42-11ea-887b-d8810560e3e9.png
[light+ new]: https://user-images.githubusercontent.com/3058150/78995569-f2377480-7b42-11ea-894a-3033d25ecb3f.png
[remedy new]: https://user-images.githubusercontent.com/3058150/78995572-f2377480-7b42-11ea-8eac-c7cfad58173d.png
[hc new]: https://user-images.githubusercontent.com/3058150/78995573-f368a180-7b42-11ea-9593-3d4c045f5820.png

robertrossmann commented:
I just noticed that the variable names in the Variables sub-view have different colour in dark variants in this PR. This was unintentional and just force-pushed a fix for that. I won't update the screens, though - it's too much work. 😇

isidorn commented on src/vs/workbench/contrib/debug/browser/debug.contribution.ts:
```diff
@@ -586,3 +587,49 @@ if (isMacintosh) {
+
+const debugTokenExpressionName = registerColor('debugTokenExpression.name', { dark: '#c586c0', light: '#9b46b0', hc: foreground }, 'Foreground color for the token names shown in the debug panel\'s Variables viewlet.');
+const debugTokenExpressionValue = registerColor('debugTokenExpression.value', { dark: '#cccccc99', light: '#6c6c6ccc', hc: foreground }, 'Foreground color for the token values shown in the debug panel\'s Variables viewlet.');
+const debugTokenExpressionString = registerColor('debugTokenExpression.string', { dark: '#ce9178', light: '#a31515', hc: '#f48771' }, 'Foreground color for strings in the debug panel\'s Variables viewlet.');
```
Instead use of "debug panel's variables view" use "debug views". That is the correct term since this color will be used across different view - Variables, Watch, Debug Console etc...

isidorn commented on src/vs/workbench/contrib/debug/browser/debugViewlet.ts:
```diff
@@ -245,3 +246,81 @@ export class OpenDebugConsoleAction extends ToggleViewAction {
 	}
 }
+
+const debugViewExceptionLabelForeground = registerColor('debugView.exceptionLabelForeground', { dark: foreground, light: foreground, hc: foreground }, 'Foreground color for a label shown in the call stack viewlet when the debugger breaks on an exception.');
```
Instead of "call stack viewlet" please use "CALL STACK view".

isidorn commented on src/vs/workbench/contrib/debug/browser/debugViewlet.ts:
```diff
@@ -245,3 +246,81 @@ export class OpenDebugConsoleAction extends ToggleViewAction {
 		super(id, label, REPL_VIEW_ID, viewsService, viewDescriptorService, contextKeyService, layoutService, 'codicon-debug-console');
 	}
 }
+
```
I suggest to move all color registration to `debug.contribution.ts`.
The reason why is that some of the debug views can be moved from the viewlet to the panel, and it can happen that the user does not have a viewlet open and thus this code will not get loaded even though some of the views might get shown.

isidorn commented on src/vs/workbench/contrib/debug/browser/repl.ts:
```diff
@@ -811,3 +812,27 @@ export class ClearReplAction extends Action {
 function getReplView(viewsService: IViewsService): Repl | undefined {
 	return viewsService.getActiveViewWithId(REPL_VIEW_ID) as Repl ?? undefined;
 }
+
```
This makes sense to live in this file, since it can not appear anywhere else.
I also suggest to completyl remove `annotationForeground` because due to a bug we do not show it. Until we fix that I suggest to not make this themable.

robertrossmann replied:
Okay, I will remove that contribution. Should I restore the original CSS for the annotation so it is not forgotten when "the time is right"?

isidorn replied:
Yeah you can restore the original css

isidorn commented:
Did another review and left comments in the code. It would be great if you could tackle them.

Yes I meant color names, so @aeschli can verify those. And once we get his feedback and @misolori feedback I will try this out a bit more and if it looks good we can merge in next week.

I am also not a fan of `debugTokenExpression` name but we can discuss that.

Thanks!

isidorn commented on src/vs/workbench/contrib/debug/browser/repl.ts:
```diff
@@ -811,3 +812,27 @@ export class ClearReplAction extends Action {
+
+const debugConsoleWarningForeground = registerColor('debugConsole.warningForeground', { dark: '#cd9731', light: '#cd9731', hc: '#008000' }, 'Foreground color of warnings in the debug console.');
+const debugConsoleAnnotationForeground = registerColor('debugConsole.annotationForeground', { dark: '#1b80b2', light: '#007acc', hc: '#0000ff' }, 'Foreground color of annotations in the debug console.');
+
```
Actually I am not a fan of both these colors `warningForeground`. I belive the debug console should not have a special warning foreground it should just use whatever other parts of debug or vscode are using.
Can we get rid of these color contributions for now. And we can tackle the debug console warning as part of another PR

robertrossmann replied:
I restored the hard-coded colour for the REPL warnings, just like I did for the annotations. ✅

isidorn commented:
Also something to keep in mind: if we ever want to make the debug colors more inlined with the language color. For example debug while deubgging javascript might want to color strings the same way as javascript colors them. This goes a bit against that. But I think that is fine. Though @weinand might be better to decide since he is looking into potential debug / language integration.

robertrossmann commented:
@isidorn Thanks for review - all concerns have been addressed.

Regarding using language token colours for the debug views - that would be the ideal scenario, of course! Until then, however, I would love to have at least something to appease my inner artistic gods. They scream loud and shudder violently whenever I start a debugging session. 😂

isidorn commented:
@robertrossmann Great work, you have my provisional approve.
Let's just finalise the naming next week.

In the future if we get nice coloring from languages then we can deprecate some of these colors.

aeschli commented:
Color names look good to me!

isidorn commented on src/vs/workbench/contrib/debug/browser/debug.contribution.ts:
```diff
@@ -586,3 +587,124 @@ if (isMacintosh) {
+const debugTokenExpressionName = registerColor('debugTokenExpression.name', { dark: '#c586c0', light: '#9b46b0', hc: foreground }, 'Foreground color for the token names shown in the debug views (ie. the Variables or Watch view).');
+const debugTokenExpressionValue = registerColor('debugTokenExpression.value', { dark: '#cccccc99', light: '#6c6c6ccc', hc: foreground }, 'Foreground color for the token values shown in the debug views (ie. the Variables or Watch view).');
+const debugTokenExpressionString = registerColor('debugTokenExpression.string', { dark: '#ce9178', light: '#a31515', hc: '#f48771' }, 'Foreground color for strings in the debug views (ie. the Variables or Watch view).');
+const debugTokenExpressionBoolean = registerColor('debugTokenExpression.boolean', { dark: '#4e94ce', light: '#0000ff', hc: '#75bdfe' }, 'Foreground color for strings in the debug views (ie. the Variables or Watch view).');
```
This should be "boolean", not "strings" in the description

robertrossmann replied:
Oops. 😇 Fixed, thanks!

isidorn commented:
@robertrossmann I found ony copy paste error in the description. Once you tackle that we can merge this in.
@misolori maybe we should look into changing some of the built in themes to customise some of these new colors. What dp you think?

robertrossmann commented:
@isidorn Thanks, I fixed the typo, squashed the review commits and rebased on current master. Excited to have this merged! 🎉 🍻

isidorn commented:
Great work, merging in. Thank you very much! ☀️

isidorn commented:
A bit of follow up polish so debug hover also uses these new colors 5a98e1353f
Found by @bpasero thanks!

robertrossmann commented:
Thanks @isidorn, I would have sworn that I tested the hovers as well! 😱 👍 ❤️

bpasero commented:
@robertohuertasm thanks ton for pushing this 👍

isidorn commented:
@robertrossmann not sure if you are interested but in case you would like these colors being used in our default themes we are open for a PR that tackles that. Of course if @misolori agrees.

robertrossmann commented:
@isidorn I think I can contribute to some themes but I am not sure if I can work on all the themes that ship with VS Code by default. 😢 Perhaps the more colourful ones would benefit the most from custom theming (Kimbie, Red, Solarized) ? The default colours work relatively well for the other themes.

@misolori Let us know what you think, we can perhaps split the work or do it incrementally as time permits. 💪

isidorn commented:
Yeah, sounds good. I definetly agree that only the more colorful one should use the new colors.
Thanks!

misolori commented:
@robertrossmann I agree that only the other non-default themes should start using the new color tokens. Feel free to ping me in your PR if you want any feedback on your proposals. Thanks again for doing this work 💪!
